activerecord/lib/active_record/associations/builder/belongs_to.rb

```diff
@@ -110,5 +110,24 @@ def self.add_destroy_callbacks(model, reflection)
       name = reflection.name
       model.after_destroy lambda { |o| o.association(name).handle_dependency }
     end
+
+    def self.define_validations(model, reflection)
+      if reflection.options.key?(:required)
+        ActiveSupport::Deprecation.warn('`required` option for belongs_to relation is deprecated and replaced with `optional` option.')
```

Comment:
What about when `required: false` is specified explicitly? Maybe throw deprecation warning only for `false` value?